Ignore extra packages in Gazelle

jayconrod opened the pull request:
If Gazelle finds .go files in a directory with multiple distinct
package declarations, it will now only generate rules to build .go
files where the package declaration matches the directory name. If
there are multiple packages, and none match the directory name,
Gazelle will report a build.MultiplePackageError, as bove. If is a
single package that doesn't match the directory name, Gazelle will
still generate rules for it.

This change helps address #339. When we are no longer applying build
constraints by default, multiple packages are much more common; many
files are hidden with "// +build ignore".

pmbethe09 commented on go/tools/gazelle/packages/walk.go:
```diff
@@ -52,3 +66,94 @@ func Walk(bctx build.Context, root string, f WalkFunc) error {
+				packageName = packageName[:len(packageName)-len("_test")]
+			}
+			packageGoFiles[packageName] = append(packageGoFiles[packageName], file)
+		} else {
```
invert this and continue
if !isGo {
  otherFiles = append(otherFiles, file)
  continue
}

pmbethe09 commented on go/tools/gazelle/packages/walk.go:
```diff
@@ -52,3 +66,94 @@ func Walk(bctx build.Context, root string, f WalkFunc) error {
+		ext := path.Ext(name)
+		isGo := ext == ".go"
+
+		var packageName string
```
remove and just use := later

pmbethe09 commented on go/tools/gazelle/packages/walk.go:
```diff
@@ -52,3 +66,94 @@ func Walk(bctx build.Context, root string, f WalkFunc) error {
+	if pname == "." || pname == "/" {
+		// We'll only use this name at the root of the filesystem.
+		return "unnamed"
+	} else {
```
drop else wrapper

pmbethe09 commented on go/tools/gazelle/packages/walk.go:
```diff
@@ -52,3 +66,94 @@ func Walk(bctx build.Context, root string, f WalkFunc) error {
+		return "", &build.NoGoError{Dir: dir}
+	} else if len(packageNames) == 1 {
+		return packageNames[0], nil
+	} else {
```
drop else wrapper

pmbethe09 commented on go/tools/gazelle/packages/walk_test.go:
```diff
@@ -31,78 +31,162 @@ func tempDir() (string, error) {
+	for _, f := range files {
+		path := filepath.Join(dir, f.path)
+		if strings.HasSuffix(f.path, "/") {
+			err = os.MkdirAll(path, 0700)
```
push into if, use :=

pmbethe09 commented on go/tools/gazelle/packages/walk_test.go:
```diff
@@ -31,78 +31,162 @@ func tempDir() (string, error) {
+			err = os.MkdirAll(path, 0700)
+			if err != nil {
+				return
+			}
```
add continue, drop else wrapper

pmbethe09 commented on go/tools/gazelle/packages/walk_test.go:
```diff
@@ -31,78 +31,162 @@ func tempDir() (string, error) {
+				return
+			}
+		} else {
+			err = os.MkdirAll(filepath.Dir(path), 0700)
```
push into if, use :=

pmbethe09 commented on go/tools/gazelle/packages/walk_test.go:
```diff
@@ -31,78 +31,162 @@ func tempDir() (string, error) {
+			if err != nil {
+				return
+			}
+			err = ioutil.WriteFile(path, []byte(f.content), 0600)
```
push into if, use :=

pmbethe09 commented on go/tools/gazelle/packages/walk_test.go:
```diff
@@ -31,78 +31,162 @@ func tempDir() (string, error) {
+		} else {
+			err = os.MkdirAll(filepath.Dir(path), 0700)
+			if err != nil {
+				return
```
log this error? (and in other places)

pmbethe09 commented on go/tools/gazelle/packages/walk_test.go:
```diff
@@ -31,78 +31,162 @@ func tempDir() (string, error) {
 		}
-		n++
+	}
+	return
```
not needed?

jayconrod commented:
@pmbethe09 Thanks for reviewing. I think I've addressed all your comments, but I can't see them inline in the code after force-pushing... sorry about that. I'll just add commits in the future.

The Travis tests were failing due to me incorrectly recognizing the "documentation" package, which go/build ignores as a special case. I've added a similar special case, and I've fixed the way I was constructing `build.MultiplePackageError`. This required changing the signature of `selectPackageName`.

PTAL.

pmbethe09 commented on go/tools/gazelle/packages/walk.go:
```diff
@@ -52,3 +66,103 @@ func Walk(bctx build.Context, root string, f WalkFunc) error {
+			packageName = name
+		}
+		return packageName, nil
+	} else {
```
remove 'else {' wrapper.

jayconrod replied:
Done. Also dropped the else above.

pmbethe09 commented on go/tools/gazelle/packages/walk.go:
```diff
@@ -52,3 +66,103 @@ func Walk(bctx build.Context, root string, f WalkFunc) error {
+			// Error() method expects these lists to be the same length. File
+			// lists are guaranteed to be non-empty.
+			err.Packages = append(err.Packages, name)
+			err.Files = append(err.Files, files[0].Name())
```
are you sure files[0] exists?

jayconrod replied:
I've updated the comment a bit to explain. `findPackageFiles` will only create an entry in `packageGoFiles` when it finds a .go file. So the list should never be empty.

pmbethe09 commented on go/tools/gazelle/packages/walk.go:
```diff
@@ -52,3 +66,103 @@ func Walk(bctx build.Context, root string, f WalkFunc) error {
+func findPackageFiles(dir string) (packageGoFiles map[string][]os.FileInfo, otherFiles []os.FileInfo, err error) {
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return
```
reads cleaner as
return nil, nil, err

jayconrod replied:
Done

pmbethe09 commented on go/tools/gazelle/packages/walk.go:
```diff
@@ -52,3 +66,103 @@ func Walk(bctx build.Context, root string, f WalkFunc) error {
+		}
+		packageGoFiles[packageName] = append(packageGoFiles[packageName], file)
+	}
+	return
```
easier to read if this is return packageGoFiles, otherFiles, nil

jayconrod replied:
Done
